packages/contracts/contracts/maci/Tally.sol

```diff
@@ -0,0 +1,321 @@
+      uint256 amount = amounts[index];
+      address receiver = receivers[index];
+
+      if (amount > 0) {
+        token.safeTransfer(receiver, amount);
+      }
+
+      unchecked {
+        index++;
+      }
+    }
+  }
+
+  /// @inheritdoc TallyBase
+  function addTallyResults(
+    uint256[] calldata voteOptionIndices,
+    uint256[] calldata results,
+    uint256[][][] calldata tallyResultProofs,
+    uint256 tallyResultSalt,
+    uint256 spentVoiceCreditsHashes,
+    uint256 perVOSpentVoiceCreditsHashes
+  ) public override isInitialized whenNotPaused afterVotingDeadline onlyOwner {
+    if (recipientCount == 0) {
+      recipientCount = registry.recipientCount();
+    }
+
+    if (recipientCount == totalTallyResults) {
+      revert TooManyResults();
+    }
+
+    super.addTallyResults(
+      voteOptionIndices,
+      results,
+      tallyResultProofs,
+      tallyResultSalt,
+      spentVoiceCreditsHashes,
+      perVOSpentVoiceCreditsHashes
+    );
+
+    if (recipientCount < totalTallyResults) {
+      revert TooManyResults();
+    }
+  }
+
+  /// @inheritdoc TallyBase
+  function addTallyResult(
+    uint256 voteOptionIndex,
+    uint256 tallyResult,
+    uint256[][] calldata tallyResultProof,
+    uint256 tallyResultSalt,
+    uint256 spentVoiceCreditsHash,
+    uint256 perVOSpentVoiceCreditsHash,
+    uint8 voteOptionTreeDepth
+  ) internal override {
+    super.addTallyResult(
+      voteOptionIndex,
+      tallyResult,
+      tallyResultProof,
+      tallyResultSalt,
+      spentVoiceCreditsHash,
+      perVOSpentVoiceCreditsHash,
+      voteOptionTreeDepth
+    );
+
+    totalVotesSquares += tallyResult ** 2;
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function claim(
+    IPayoutStrategy.Claim calldata params
+  ) public override isInitialized whenNotPaused afterVotingDeadline {
+    uint256 amount = getAllocatedAmount(params.voiceCreditsPerOption, params.totalSpent, params.tallyResult);
+    totalAmount -= amount;
+
+    bool isValid = verifyTallyResult(
+      params.index,
+      params.tallyResult,
+      params.tallyResultProof,
+      params.tallyResultSalt,
+      params.voteOptionTreeDepth,
+      params.spentVoiceCreditsHash,
+      params.perVOSpentVoiceCreditsHash
+    );
+
+    if (!isValid) {
+      revert InvalidTallyVotesProof();
+    }
+
+    IRecipientRegistry.Recipient memory recipient = registry.getRecipient(params.index);
+
+    token.safeTransfer(recipient.recipient, amount);
+  }
+
+  /// @notice Get allocated token amount (without verification).
+  /// @param voiceCreditsPerOption The voice credit options received for recipient
+  /// @param totalSpent The total amount of voice credits spent
+  /// @param tallyResult The tally result for recipient
+  function getAllocatedAmount(
+    uint256 voiceCreditsPerOption,
+    uint256 totalSpent,
+    uint256 tallyResult
+  ) internal view returns (uint256) {
+    uint256 alpha = calculateAlpha(totalSpent);
```

Comment:
do you think it's best this is calculated only once so that we make claims cheaper?

Reply:
It won't be possible with current design because users can call `getAllocatedAmount` and check the recipient potential amount.
In case if we want to save this alpha somewhere, we need to update prove onchain logic (update it in claim won't work because if there are no claims, so alpha will be zero).

Reply:
why would alpha be zero with no claims?

Reply:
Because `totalVotesSquares` will be zero and we won't be able to calculate alpha, so it will be a default value

Reply:
doesn't totalVotesSquares comes from addTallyResult? claims should not happen until all batches have been processed with addTallyResults

Reply:
Right, `totalVotesSquares` will be available after `addTallyResults` call and `claim` will revert if there are no `totalVotesSquares`

Reply:
won't that also be a problem if the coordinator is not able to upload all results in one batch (due to block gas limit) and one of the projects submits a transaction to claim before the second batch of results are published?

Reply:
I don't think so, because there is isTallied check for claim. I think I can add this check only for calculations if users want to use only view functions.